explore/src/main/scala/explore/programs/ProgramsPopup.scala

```diff
@@ -0,0 +1,170 @@
+
+  protected def selectProgram(
+    isRequired: Boolean,
+    isOpen:     ReuseView[Boolean],
+    undoStacks: ReuseView[ModelUndoStacks[IO]],
+    programId:  Program.Id
+  )(implicit
+    ctx:        AppContextIO
+  ): Callback =
+    isOpen.set(false) >>
+      undoStacks.set(ModelUndoStacks[IO]()) >>
+      (if (isRequired) ctx.replacePage(AppTab.Overview, programId, none, none)
+       else ctx.pushPage(AppTab.Overview, programId, none, none))
+
+  protected def onNewData(
+    isOpen:     ReuseView[Boolean],
+    isRequired: Boolean,
+    programs:   List[ProgramInfo]
+  )(implicit
+    ctx:        AppContextIO
+  ): IO[Unit] =
+    (isRequired, programs) match {
+      case (false, _)          => IO.unit
+      case (true, head :: Nil) => ctx.replacePage(AppTab.Overview, head.id, none, none).to[IO]
+      case _                   => isOpen.set(true).to[IO]
+    }
+
+  protected def showModal(
+    props:       Props,
+    isOpen:      ReuseView[Boolean],
+    adding:      ReuseView[Boolean],
+    showDeleted: ReuseView[Boolean],
+    isRequired:  Boolean,
+    programs:    ReuseView[List[ProgramInfo]]
+  ): VdomNode = {
+    implicit val ctx = props.ctx
+
+    val actions =
+      if (isRequired) List.empty
+      else
+        List(
+          Button(size = Small, icon = true)(Icons.Close, "Cancel")(^.tpe := "button",
+                                                                   ^.key := "input-cancel"
+          )
+        )
+
+    React.Fragment(
+      Modal(
+        clazz = ExploreStyles.ProgramsPopup,
+        actions = actions,
+        centered = false,
+        open = isOpen.get,
+        closeOnDimmerClick = !isRequired,
+        closeOnEscape = !isRequired,
+        closeIcon =
+          if (isRequired) js.undefined
+          else Icons.Close.clazz(ExploreStyles.ModalCloseButton),
+        dimmer = Dimmer.Blurring,
+        size = ModalSize.Small,
+        onClose = isOpen.set(false),
+        header = ModalHeader(
+          "Programs"
+        ),
+        content = ModalContent(
+          ProgramTable(
+            props.currentProgramId,
+            programs,
+            showDeleted.get,
+            selectProgram = Reuse.currying(isRequired, isOpen, props.undoStacks).in(selectProgram _)
+          ),
+          <.div(
+            Button(
+              clazz = ExploreStyles.ProgramAdd,
+              // size = Small,
+              compact = true,
+              icon = Icons.New,
+              content = "Program",
+              disabled = adding.get,
+              loading = adding.get,
+              onClick = addProgram(programs, adding).runAsync
+            ),
+            Checkbox(label = "Show deleted",
+                     checked = showDeleted.get,
+                     onChange = showDeleted.set _
+            )
+          )
+        )
+      )
+    )
+  }
+
+  protected val component = ScalaFnComponent
+    .withHooks[Props]
+    .useStateViewWithReuse(false) // isOpen
```

Comment:
I hope when we have opaque types we could make these so it is clearer which one is doing what, but that's unrelated to this PR

Reply:
Right. There are a lot of Booleans. ☹️